packages/trader/src/RangeTrader.ts

```diff
@@ -0,0 +1,183 @@
+const { ConvertDecimals, createFormatFunction, createObjectFromDefaultProps } = require("@uma/common");
+import assert from "assert";
+
+export class RangeTrader {
+  readonly logger: any;
+  readonly web3: any;
+  readonly tokenPriceFeed: any;
+  readonly referencePriceFeed: any;
+  readonly normalizePriceFeedDecimals: any;
+  readonly formatDecimalString: any;
+  readonly exchangeAdapter: any;
+
+  // TODO: is there any way to not have to declare these as part of the type?
+  readonly BN: any;
+  readonly toBN: any;
+  readonly toWei: any;
+  readonly fromWei: any;
+
+  readonly tradeExecutionThreshold: any;
+  readonly targetPriceSpread: any;
+  readonly fixedPointAdjustment: any;
```

Comment:
this means no longer creating a class for the `RangeTrader` and just baking everything into one nested function? that's a really interesting pattern, I have not seen that before. What are the implications seeing we no longer have a "class" object. I mean, it amounts to effectively the same thing I guess but what are the tradeoffs with this pattern?

Reply:
Yes the drawback is there are not really any classes, though with typescript you can create interfaces to check this.  This class though in particular is not meant to really be extended or changed, so I dont think you lose much in this case. Ultimately this is more of a style thing and a way to get around using `this` and having all this extra ceremony around storing class variables and setting visibility, functionally is 99% equivalent.  It pains me to see so much extra code that could be avoided.

Reply:
Yeah, it's a style thing, and I totally get the anti-verbosity argument.

I _personally_ don't mind (and even prefer) the verbosity of explicitly declaring instance/class variables. @daywiss, you and I have had really good discussions on this before, but I think using classes makes the _scoping_ of data very clear via declarations, especially in typescript where they are declared up-front. I feel like closures kind of mask scoping, and they make object lifetimes and data availability a little more difficult to reason about.

One other big advantage of classes IMO is tooling. Classes are a much more common pattern and my guess is that typescript type inference, IDE tooling, etc _just works better_ with classes than they do with this closure-class pattern.

The last point I'll make in favor of classes is that variable declaration is a little verbose up-front, but it's really not _that_ bad. My guess is that VSCode has some feature where it can add them for you.

Reply:
thanks for the input above guys ^^. I'd say that for now we leave it as a class (even though it is ugly and verbose, with you on that @daywiss ) and then we can converge on a _common_ styling pattern in the future. we can always refactor this later if we choose but for now consistency is king 👑